src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -3,45 +3,205 @@
+                remove(index);
+                return true;
+            }
+            index++;
+            node = node.next;
+        }
         return false;
     }
 
     @Override
     public int size() {
-        return 0;
+        return size;
     }
 
     @Override
     public boolean isEmpty() {
-        return false;
+        return size == 0;
+    }
+
+    private T removeLast() {
+        Node<T> nodeToRemove = endNode;
+        unlink(endNode);
+        size--;
+        return nodeToRemove.data;
+    }
+
+    private T removeFirst() {
+        Node<T> nodeToRemove = startNode;
+        unlink(startNode);
+        size--;
+        return nodeToRemove.data;
+    }
+
+    private void unlink(Node<T> node) {
+        if (node == startNode) {
+            if (startNode.next != null) {
+                startNode = node.next;
+            }
+            return;
+        }
+        if (node == endNode) {
+            endNode = node.previous;
+            endNode.next = null;
+            return;
+        }
+        node.previous.next = node.next;
+        node.next.previous = node.previous;
+    }
+
+    private T removeByIndex(int index) {
+        Node<T> node = findByIndex(index);
+        unlink(node);
+        size--;
+        return node.data;
+    }
+
+    private void linkLast(T value) {
+        if (size == 0) {
+            startNode = new Node<>(value, endNode, null);
+        } else if (size == 1) {
+            endNode = new Node<>(value, null, startNode);
+        } else {
+            endNode = new Node<>(value, null, endNode);
+        }
+        size++;
+    }
+
+    private void linkFirst(T value) {
+        if (size == 0) {
+            startNode = new Node<>(value, endNode, null);
+        } else {
+            startNode = new Node<>(value, startNode, null);
+        }
+        size++;
+    }
+
+    private void linkAtMiddle(T value, int index) {
+        Node<T> existingNode = findByIndex(index);
+        new Node<>(value,existingNode, existingNode.previous);
+        size++;
+    }
+
+    private Node<T> findByIndex(int index) {
+        indexValidation(index);
+        if (index >= 0 && index < 2 && size < 2) {
+            return index == 0 ? startNode : endNode;
+        }
+        if (index - size / 2 <= 0) {
+            return findNodeFromStart(index);
+        } else {
+            return findNodeFromEnd(index);
+        }
+    }
+
+    private Node<T> findNodeFromStart(int index) {
+        Node<T> x = startNode;
+        int currentIndex = 0;
+        while (index != currentIndex) {
+            x = x.next;
+            currentIndex++;
+        }
+        return x;
+    }
+
+    private Node<T> findNodeFromEnd(int index) {
+        Node<T> x = endNode;
+        int currentIndex = size - 1;
+        while (index != currentIndex) {
+            x = x.previous;
+            currentIndex--;
+        }
+        return x;
+    }
+
+    private void indexValidation(int index) {
```

Comment:
method names should be verbs
```suggestion
    private void validateIndex(int index) {
```